File: src/leetcode/backtrack/51.N皇后.ts
/*
 * @lc app=leetcode id=51 lang=typescript
 *
 * [51] N-Queens
 */

// @lc code=start
function solveNQueens(n: number): string[][] {
  const board = Array(n)
    .fill(0)
    .map(() => '.'.repeat(n));
  let res = [];

  backtrack(0);

  return res;

  function backtrack(r) {
    if (r === n) {
      res.push([...board]);
      return;
    }

    for (let c = 0; c < n; c++) {
      if (!isValid(r, c)) continue;

      board[r] = board[r].substring(0, c) + 'Q' + board[r].substring(c + 1);
      backtrack(r + 1);
      board[r] = board[r].substring(0, c) + '.' + board[r].substring(c + 1);
    }
  }

  function isValid(r, c) {
    // 检查列
    for (let i = 0; i < r; i++) {
      if (board[i][c] === 'Q') return false;
    }

    // 检查左上对角线
    for (let i = r - 1, j = c - 1; i >= 0 && j >= 0; i--, j--) {
      if (board[i][j] === 'Q') return false;
    }

    // 检查右上对角线
    for (let i = r - 1, j = c + 1; i >= 0 && j < n; i--, j++) {
      if (board[i][j] === 'Q') return false;
    }

    return true;
  }
}

if (require.main === module) {
  let res = solveNQueens(4);

  for (const board of res) {
    for (const row of board) {
      console.log(row.split('').join(' '));
    }
    console.log();
  }
}
// @lc code=end

export {};

File: src/leetcode/backtrack/51.N 皇后.py
#
# @lc app=leetcode.cn id=51 lang=python3
# @lcpr version=30104
#
# [51] N 皇后
#

# @lc code=start


from typing import List


class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        board = [[0 for _ in range(n)] for _ in range(n)]
        res = []

        def valid(board, row, col):
            for i in range(row):
                if board[i][col] == 1:
                    return False

            for r, c in zip(range(row - 1, -1, -1), range(col - 1, -1, -1)):
                if board[r][c] == 1:
                    return False

            for r, c in zip(range(row - 1, -1, -1), range(col + 1, n, 1)):
                if board[r][c] == 1:
                    return False

            return True

        def backtrack(board: list[list[int]], row: int):
            if row == n:
                solution = []
                for r in board:
                    solution.append("".join("Q" if num == 1 else "." for num in r))
                res.append(solution)
                return

            for c in range(n):
                if not valid(board, row, c):
                    continue
                board[row][c] = 1
                backtrack(board, row + 1)
                board[row][c] = 0

        backtrack(board, 0)
        return res


# @lc code=end


#
# @lcpr case=start
# 4\n
# @lcpr case=end

# @lcpr case=start
# 1\n
# @lcpr case=end

#
